tracing-attributes/src/lib.rs

```diff
@@ -254,30 +255,51 @@ pub fn instrument(
 ) -> proc_macro::TokenStream {
-    let input: ItemFn = syn::parse_macro_input!(item as ItemFn);
+    let input = syn::parse_macro_input!(item as ItemFn);
     let args = syn::parse_macro_input!(args as InstrumentArgs);
 
     let instrumented_function_name = input.sig.ident.to_string();
 
-    // check for async_trait-like patterns in the block and wrap the
-    // internal function with Instrument instead of wrapping the
-    // async_trait generated wrapper
+    // check for async_trait-like patterns in the block, and instrument
+    // the future instead of the wrapper
     if let Some(internal_fun) = get_async_trait_info(&input.block, input.sig.asyncness.is_some()) {
         // let's rewrite some statements!
-        let mut stmts: Vec<Stmt> = input.block.stmts.to_vec();
-        for stmt in &mut stmts {
-            if let Stmt::Item(Item::Fn(fun)) = stmt {
-                // instrument the function if we considered it as the one we truly want to trace
-                if fun.sig.ident == internal_fun.name {
-                    *stmt = syn::parse2(gen_body(
-                        fun,
-                        args,
-                        instrumented_function_name,
-                        Some(internal_fun),
-                    ))
-                    .unwrap();
-                    break;
+        let mut out_stmts = Vec::with_capacity(input.block.stmts.len());
+        for stmt in &input.block.stmts {
+            if stmt == internal_fun.source_stmt {
+                match internal_fun.kind {
+                    // async-trait <= 0.1.43
+                    AsyncTraitKind::Function(fun) => {
+                        out_stmts.push(gen_function(
+                            fun,
+                            args,
+                            instrumented_function_name,
+                            Vec::new(),
+                            internal_fun.self_type,
+                        ));
+                    }
+                    // async-trait >= 0.1.44
+                    AsyncTraitKind::Async(async_expr) => {
+                        // fallback if we couldn't find the '__async_trait' binding, might be
+                        // useful for crates exhibiting the same behaviors as async-trait
+                        let instrumented_block = gen_block(
+                            &async_expr.block,
+                            &input.sig.inputs,
+                            true,
+                            args,
+                            instrumented_function_name,
+                            Vec::new(),
+                            None,
+                        );
+                        let async_attrs = &async_expr.attrs;
+                        // we use liballoc to allow for no_std environments
+                        out_stmts.push(quote! {
+                            extern crate alloc;
+                            alloc::boxed::Box::pin(#(#async_attrs) * async move { #instrumented_block })
```

Comment:
hmm, i'm not actually sure if generating an `extern crate` statement for `alloc` is actually the best thing. looking at `async-trait`, they just generate an un-prefixed `Box::pin`: https://github.com/dtolnay/async-trait/blob/a4aab7b9285e4912d16f1f4733cfb99fc8aa37da/src/expand.rs#L385-L387

maybe this is more correct, since it doesn't silently add `alloc` to no-std projects which don't depend on `alloc`? it won't compile out of the box on `core` + `alloc` projects, but if users add a `use alloc::boxed::Box` at the top of the file, it will; otherwise, it'll use `Box` from the prelude when libstd is there.

i know i said otherwise earlier, but after giving it some more thought, I think maybe it's best to do "whatever `async-trait` does" in this case. maybe @dtolnay has input?

Reply:
On that note, i'm trying to test in a `no_std` environement, but it's not taht easy considering the fact the the support module is using String and similar standard library features. Would you be interested in a PR that try to adress this? (Not really commiting to anything here, but I won't start working on this if this isn't interesting for the project)